### ISSUE TITLE
[PoSMinting] reserved balance is not implemented properly

### ISSUE THREAD

**Issue by noescape00** [stratisproject/StratisBitcoinFullNode#754]
In FindCoinsForStaking (if PR #753 gets accepted referenced function name is GetTxesSuitableForStaking) function there is an input value 'long maxValue' with description 'Maximal amount of coins that can be used for staking.' but it actually used for limiting max TxOut value for each transaction that participates in staking separately and not in total how it was supposed to. 
So for example if you set maxValue to 10, but there are 5 transactions and each is <=10 maxValue will be ignored and nothing will be reserved from staking .

**Comment by dangershony** [comment]
Hmm interesting, the point of this is to always have some coins available for spending so you are right that condition may not be fulfilled, your example of 5 trx and each finds a stake one after the other will break this requirement.

**Comment by dangershony** [comment]
I think `stakeTx.TxOut.Value` is in satoshi, if you look at the Money class everything is converted to satoshi

**Comment by noescape00** [comment]
I guess it's not architecturally even possible to prevent a certain amount from staking, only transaction(s). Unless I guess we create transaction to divide existing one into 2 of certain sizes (I mean if I have only one UTXO with a size of 10 strat and I want to reserve 3 strat I can send 3 strat to myself so I'd have 2 UTXOs with a size of 3 and 7 strat and I'd stake one with 7 strat). But it's probably not a good idea since I can end up paying more fees just to reserve a certain amount than I'd get from staking. 

>I think stakeTx.TxOut.Value is in satoshi, if you look at the Money class everything is converted to satoshi
Right, my bad.

**Comment by dangershony** [comment]
It is complicated yes if you propose an algo to make this more efficient I am fine with it, but its not high priority in my opinion.

**Comment by noescape00** [comment]
I think it will be a good idea ether to delete possibility to set reserve amount entirely or to change it to MinReserveAmount which will reserve at least provided amount.

**Comment by dangershony** [comment]
Ah wait i might be wrong with my comment of the reason for this flag.
Let's go to stratis X see if there is a bit more info on that.

**Comment by noescape00** [comment]
>Let's go to stratis X see if there is a bit more info on that.

In StratisX=>wallet.cpp:

Line 1620:
`SelectCoinsForStaking(nBalance - nReserveBalance, txNew.nTime, setCoins, nValueIn))`

Line 1357:
`
bool CWallet::SelectCoinsForStaking(int64_t nTargetValue, unsigned int nSpendTime, set<pair<const CWalletTx*,unsigned int> >& setCoinsRet, int64_t& nValueRet) const
{
    ....
    BOOST_FOREACH(COutput output, vCoins)
    {
        // Stop if we've chosen enough inputs
        if (nValueRet >= nTargetValue)
            break;

        int64_t n = pcoin->vout[i].nValue;
        ...
        if (n >= nTargetValue)
        {
            // If input value is greater or equal to target then simply insert
            //    it into the current subset and exit
            setCoinsRet.insert(coin.second);
            nValueRet += coin.first;
            break;
        }
        else if (n < nTargetValue + CENT)
        {
            setCoinsRet.insert(coin.second);
            nValueRet += coin.first;
        }
       .....
`

So basically it will get enough UTXO's to reach target balance and leave the rest untouched.
